core/src/main/java/org/geysermc/geyser/registry/populator/CustomItemRegistryPopulator.java

```diff
@@ -0,0 +1,328 @@
+            }
+
+            int javaCustomItemId = javaItem.getJavaId();
+            int customItemId = itemMappings.getItems().size() + (customItems * protocolVersions.size()) + addNum;
+
+            ItemMapping customItemMapping = ItemMapping.builder()
+                    .javaIdentifier(baseItem)
+                    .bedrockIdentifier(customItemName)
+                    .javaId(javaCustomItemId)
+                    .bedrockId(customItemId)
+                    .bedrockData(0)
+                    .bedrockBlockId(-1)
+                    .stackSize(javaItem.getStackSize())
+                    .build();
+
+            StartGamePacket.ItemEntry startGamePacketItemEntry = new StartGamePacket.ItemEntry(customItemName, (short) customItemId, true);
+
+            NbtMapBuilder builder = createItemData(customItemData, javaItem, customItemName, customItemId, OptionalInt.empty(), null, false, javaItem.isTool());
+
+            if (customItemData.customItemOptions().hasCustomItemOptions()) {
+                javaItem.getCustomItemOptions().put(customItemData.customItemOptions(), customItemId);
+            } else {
+                GeyserImpl.getInstance().getLogger().warning("The custom item " + customItemData.name() + " has no registration types");
+            }
+
+            ComponentItemData componentItemData = new ComponentItemData(customItemName, builder.build());
+            returnData.addMapping(protocolVersion, new GeyserCustomMappingData.Mapping(componentItemData, customItemMapping, startGamePacketItemEntry, customItemName, customItemId));
+            itemMappings.getCustomIdMappings().put(customItemId, customItemName);
+        }
+
+        return returnData;
+    }
+
+    public static Int2ObjectMap<ComponentItemData> populateRegistry(NonVanillaCustomItemData customItemData, int customItems) {
+        Int2ObjectMap<ComponentItemData> componentItemDataMap = new Int2ObjectOpenHashMap<>();
+        int addNum = 0;
+
+        for (Int2ObjectMap.Entry<ItemMappings> entry : Registries.ITEMS.get().int2ObjectEntrySet()) {
+            addNum++;
+
+            int protocolVersion = entry.getIntKey();
+            ItemMappings itemMappings = entry.getValue();
+
+            boolean exists = false;
+            for (ItemMapping mapping : itemMappings.getItems().values()) {
+                if (mapping.getJavaIdentifier().equals(customItemData.identifier()) || mapping.getBedrockIdentifier().equals(customItemData.identifier()) || mapping.getJavaId() == customItemData.javaId()) {
+                    exists = true;
+                    break;
+                }
+            }
+            if (exists) {
+                GeyserImpl.getInstance().getLogger().error("The custom item " + customItemData.identifier() + " is overwriting an existing item in protocol version " + protocolVersion +"! Could not register it.");
+                continue;
+            }
```

Comment:
It'd be a bit ugly but I'd probably do the log and continue stuff inside the for loop.
This would require a label in the parent for loop.